_posts/2019-07-30-microfrontends-holidaycheck-part2.md

```diff
@@ -0,0 +1,168 @@
+we need help, report bugs or suggest features, I strongly recommend everyone to try their service.
+
+<figure>
+    <figcaption>Overview of a test run with Calibre</figcaption>
+    <img src="/img/posts/2019-07-30-microfrontends-holidaycheck-part2/calibre.png" alt="calibre-overview" />
+</figure>
+
+## Analysing
+For deeper analysis we use Google Lighthouse together with the various capabilities of the Chrome Dev Tools.
+Explaining the Chrome Dev Tools would exceed the scope of the article. I recommend everyone to get familiar
+with the Dev Tools. When it comes to speed especially the Network, Performance and Audits tabs are relevant.
+
+## Bundle Size
+In the today's JS ecosystem it is quite a challenge to build an App with a modern component based approach
+within 32 kB. Considering React + React-Dom have already 110 kB minified.
+
+That's why we decided to use Preact + Unistore, which together are roughly 10 kB minified. Preact has the same
+API as React with a similar feature set (besides synthetic events), but is *a lot* smaller. Unistore is the
+complementary tiny state management solution.
+
+It still happened regularly that we hit the 32kB for our main bundle and our CI pipeline gently reminded
+us with a red build. To debug what takes up the budget the Webpack Bundle Analyzer plugin is really great.
+It generates an interactive tree map which let's you easily find the big chunks.
+
+<figure>
+    <figcaption>Tree Map generated by Webpack Bundle Analyzer</figcaption>
+    <img src="/img/posts/2019-07-30-microfrontends-holidaycheck-part2/analyze-bundle.png" alt="bundle-analyzer-tree-map" />
+</figure>
+
+### Bundle Splitting
+Once we found the problematic parts we needed to find a solution. The biggest effect had **bundle splitting**.
+
+On the first impression we shall deliver only the Javascript which is absolutely necessary for the current
+view. Everything else can be loaded asynchronously on demand. Possible split points are things which are needed
+after an interaction, like opening the menu or rendering of search results. Also parts of the code which 
+are needed just by portion of our users, like features which are only available once you are logged in, are 
+a great possibility to add a split point.
+
+Adding a split point when using webpack is quite simple by using the `import` keyword. 
+
+```js
+import(/* webpackChunkName: "desktopMenu" */ './DesktopMenu').then(menu => {
+  console.log("Menu loaded", menu)
+});
+```
+The import keyword will return a promise which completes with whatever the loaded module exports. The
+effect of this is that webpack will separate all code of the `DesktopMenu` including its dependencies into 
+a separate JS bundle which is loaded only once the import is triggered. To simplify and reduce
+boilerplate for this, React offers the [lazy](https://reactjs.org/docs/code-splitting.html#reactlazy) function. 
```

Comment:
We could add a part about the "gotchas" that we encountered along the way. Like when we had to use the "ComponentWithDefaults" strategy to avoid loading data and extra code into the main bundle.

Reply:
Hmm I think that would be too much in here, but suit better in an article which concentrates completely on bundle splitting.